### PR TITLE
embed: warns about empty hosts in advertise urls

### PR DESCRIPTION
**Pull request by gyuho**
Address https://github.com/coreos/etcd/issues/8379.

> ./bin/etcd --listen-client-urls=http://127.0.0.1:2379 --advertise-client-urls=http://$SPELEDOK:2379

> 2017-08-09 13:34:52.220844 E | etcdmain: error verifying flags, unexpected empty host (http://:2379). See 'etcd --help'.

### CODE REVIEW

**Comment by heyitsanthony** [comment]
I tried some empty host url abuse.

./etcd defaults:
```sh
$ ETCDCTL_API=3 ./bin/etcdctl --endpoints=http://:2379 endpoint status
http://:2379, 8e9e05c52164694d, 3.2.0+git, 16 kB, true, 10, 20
```

fails because of IP check:
```sh
$ ./bin/etcd --listen-client-urls=http://:2379 --advertise-client-urls=http://:2379
etcdmain: error verifying flags, expected IP in URL for binding (http://:2379). 
```

(fixed by this patch):
```sh
$ ./bin/etcd --listen-client-urls=http://0.0.0.0:2379 --advertise-client-urls=http://:2379
```

Does anyone use the empty host => any address behavior? A workaround like this seems tempting:
```sh
$ ETCDCTL_API=3 ./bin/etcdctl --endpoints=http://0.0.0.0:2379 endpoint status
http://0.0.0.0:2379, 8e9e05c52164694d, 3.2.0+git, 16 kB, true, 12, 24
```
but that only binds to any ipv4 address instead of any address.

On the other hand, empty hosts fail the `curl` sanity test:
```sh
$ curl http://:2379/
curl: (6) Could not resolve host: 
$ curl http://localhost:2379/health
{"health":true}
$ wget http://:2379/health
http://:2379/health: Invalid host name.
```

If etcd is advertising a client address curl can't parse, that must be broken. Maybe someone somewhere somehow relies on this though. Can this be a warning in 3.3 then a full error/exit in 3.4?

**Comment by gyuho** [comment]
@heyitsanthony PTAL. Now just logs warnings.

Thanks.

File: embed/config.go [hunk 1]
```diff
@@ -299,6 +299,14 @@ func (cfg *Config) Validate() error {
 	if err := checkBindURLs(cfg.ListenMetricsUrls); err != nil {
 		return err
 	}
+	if err := checkHostURLs(cfg.APUrls); err != nil {
+		// TODO: return err in v3.4
+		plog.Warning(err)
```
Comment (heyitsanthony) [inline embed/config.go]: should these say something like "advertise-peer-urls %q is deprecated (%v)" if the intent is to eventually exit out? warning with only the error doesn't seem clear enough

**Comment by gyuho** [comment]
@heyitsanthony PTAL. Now

> ./bin/etcd --listen-client-urls=http://127.0.0.1:2379 --advertise-client-urls=http://$SPELEDOK:2379
2017-08-10 12:33:58.211006 W | embed: advertise-client-urls "http://:2379" is deprecated (unexpected empty host (http://:2379))

Thanks.

**Pull request merged**